src/main/cs/RLBotDotNet/RLBotDotNet/Manager/BotManager.cs

```diff
@@ -65,18 +66,26 @@ private void RunBot(Bot bot)
         /// </summary>
         private void MainBotLoop()
         {
+            TimeSpan timerResolution = TimerResolutionInterop.CurrentResolution;
+            TimeSpan targetSleepTime = new TimeSpan(166667); // 16.6667 ms, or 60 FPS
+
+            Stopwatch stopwatch = new Stopwatch();
             while (true)
             {
-                try
-                {
-                    botRunEvent.Set();
-                    botRunEvent.Reset();
-                    Thread.Sleep(16);
-                }
-                catch (Exception e)
-                {
-                    Console.WriteLine(e.Message);
-                }
+                // Start the timer
+                stopwatch.Restart();
+
+                // Set off events that end up running the bot code later down the line
+                botRunEvent.Set();
+                botRunEvent.Reset();
+
+                // Sleep efficiently (but inaccurately) for as long as we can
+                TimeSpan maxInaccurateSleepTime = targetSleepTime - stopwatch.Elapsed - timerResolution;
+                if (maxInaccurateSleepTime > TimeSpan.Zero)
+                    Thread.Sleep(maxInaccurateSleepTime);
+
+                // Sleep the rest of the time accurately with the use of a tight loop
+                while (stopwatch.Elapsed < targetSleepTime);
```

Comment:
Can you comment on the performance implications here? I'm worried that this will consume a lot of CPU, and might impact people's ability to stream tournaments smoothly, or run other bots on the same team smoothly.

Reply:
Assuming a 0.5 ms accurate timer (which is the most common value), this while loop will run for about 3% of sleep time (with the rest handled normally by Thread.Sleep). Omitting this line will make the timings inconsistent, but ticks will never be late (only slightly too early).

Reply:
My personal preference is to omit that line and not risk unreasonable CPU utilization. Have an opinion here, Redox?

Reply:
I don't think this empty while loop will perform very efficiently. I think it's better to have slightly early timings rather than having someone's CPU be used up a lot. My opinion here is that the line should be omitted, but if I'm incorrect and it uses up a negligible (or very little) amount of CPU then it's fine to keep.

Reply:
Alright, I must agree that it's not critical for this loop to fire at precisely 60hz - it just needs to not be late, for which the while loop is not needed.